Add frontend for working with storage

romanhabibov opened the pull request:
Implement GET method and simple html frontend. Now, it is possible
to work with the storage using a browser.

Closes #10

Totktonada commented:
Please, remove the `.DS_Store` file.

How icons you added are licensed and who is their author?

Whether `index.html` is derived work (based on some other work)? If so, how it is licensed and who is the author?

LeonidVas commented on s3repo/model.py:
```diff
@@ -378,9 +388,103 @@ def delete_package(self, package):
-        NotImplementedError("get_file hasn't been implemented yet.")
+        """Get a file from S3."""
+
+        self.sync_lock.acquire()
```
This mutex is used to protect "unsync_repos".
Why do you need mutex at all?

romanhabibov replied:
I think files while downloading may be a competing stream.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,26 @@
+from flask import render_template, Response
+import os
+
+class IndexView(MethodView):
```
Why do you use inheritance from "MethodView"? Seems like the "View" class can be used here.

romanhabibov replied:
I tried, as_view() needs "MethodView" heir.

LeonidVas replied:
Sorry, but I don't understand you. Seems like the View class has the `as_view()` method (https://flask.palletsprojects.com/en/2.0.x/views/)

romanhabibov replied:
Done, but I renamed View's get() to dispatch_request() as stated in the article.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,26 @@
+from flask.views import MethodView
+from flask import render_template, Response
+import os
```
https://www.python.org/dev/peps/pep-0008/#imports (Imports should be grouped in the following order:....)

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,26 @@
+from flask.views import MethodView
+from flask import render_template, Response
+import os
+
```
Add blank line.

romanhabibov replied:
?

LeonidVas replied:
![image](https://user-images.githubusercontent.com/57662928/124025691-70f32880-d9f9-11eb-8de9-c1e45f5f2146.png)

From https://www.python.org/dev/peps/pep-0008/#imports
```
Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.
```

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,26 @@
+from flask.views import MethodView
```
Add module description.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,26 @@
+    """IndexView is just methods to display listings or download
+    files with browser.
+    """
+    def get_directory(self, subpath, files, folders):
```
I think the class should have one public method get and two private ("get_directory" and "get_file").

romanhabibov replied:
Done. See the patch about s3view.

LeonidVas reviewed (changes requested):
Hi! Thank you for the patch. I recommend you use the "flake8" tool to check your code. And read the "PEP8".
See some comments bellow:

LeonidVas commented:
Yet few comments / thoughts:
- We should not show the entire bucket, but start from the bucket/base/path
- Maybe we want to have a filter for files that shouldn't be shown (what do you think)
- Split the patch into commits

LeonidVas commented on s3repo/model.py:
```diff
@@ -378,9 +388,103 @@ def delete_package(self, package):
+        directory from S3.
+        """
+
+        paginator = self.s3_client.get_paginator('list_objects')
```
Use "list_objects_v2"

romanhabibov replied:
Done.

LeonidVas commented on s3repo/model.py:
```diff
@@ -378,9 +388,103 @@ def delete_package(self, package):
+        """
+
+        paginator = self.s3_client.get_paginator('list_objects')
+        list_parameters = {'Bucket': self.bucket.name,
```
AFAIU if you add "Delimiter": '/', then you will only get the keys inside the directory (files in the "Key" section and directories in the "CommonPrefixes" section)
```
{'ResponseMetadata': {'HTTPStatusCode': 200, 'HTTPHeaders': {'server': 'nginx/1.16.1', 'date': 'Thu, 17 Jun 2021 09:05:03 GMT', 'content-type': 'application/xml', 'content-length': '800', 'connection': 'keep-alive', 'x-req-id': 'z8q33oq5', 'x-host': 'hb-front9'}, 'RetryAttempts': 0}, 'IsTruncated': False, 'Contents': [{'Key': 'Sources (3)', 'LastModified': datetime.datetime(2021, 6, 17, 9, 4, 44, 928000, tzinfo=tzutc()), 'ETag': '"975d93622e557d183efa334049dc4ee6"', 'Size': 1580, 'StorageClass': 'STANDARD'}, {'Key': 'test-dir', 'LastModified': datetime.datetime(2021, 6, 3, 9, 15, 41, 535000, tzinfo=tzutc()), 'ETag': '"d41d8cd98f00b204e9800998ecf8427e"', 'Size': 0, 'StorageClass': 'STANDARD'}], 'Name': 'tarantool-dev', 'Prefix': '', 'Delimiter': '/', 'MaxKeys': 1000, 'CommonPrefixes': [{'Prefix': 'check/'}, {'Prefix': 'test-dir/'}], 'EncodingType': 'url', 'KeyCount': 4}
```

romanhabibov replied:
Done.

LeonidVas commented on s3repo/model.py:
```diff
@@ -378,9 +388,103 @@ def delete_package(self, package):
+                            size = fileobj.get('Size')
+
+                            # Make size human readable.
+                            human_readadble = ''
```
Move this logic into a separate private method.

LeonidVas replied:
It also looks like the transforming size to human-readable needs to be moved to the view.

romanhabibov replied:
Maybe it would be more logical to do it in the view, but it is more convenient as it is now. I'll leave it for now.

LeonidVas replied:
No, you shouldn't do that, because there is a mixture of data and the way it is displayed. I don't understand what are the difficulties to transfer this logic to the view.

romanhabibov replied:
Ok, done.

LeonidVas commented:
> > Please, remove the `.DS_Store` file.
> > How icons you added are licensed and who is their author?
> > Whether `index.html` is derived work (based on some other work)? If so, how it is licensed and who is the author?
> 
> +1
> @romanhabibov ?

I think we should use something like [this one](https://google.github.io/material-design-icons/).

LeonidVas commented on templates/index.html:
```diff
@@ -0,0 +1,74 @@
+<body>
+    <h1> Index of {{ path }} </h1>
+    <table id="indexlist">
+        <tr class="indexhead">
```
Maybe I don't understand something, but where did you define the classes "indexhead", "indexcolicon", "indexcolname" ...

romanhabibov replied:
Dropped.

LeonidVas commented on templates/index.html:
```diff
@@ -0,0 +1,74 @@
+        {% for file in files %}
+            {% if loop.index % 2 == 0 %}
+                <tr class="even">
+            {% else %}
```
Look at https://jinja.palletsprojects.com/en/3.0.x/tricks/#alternating-rows

Totktonada replied:
We can use pure CSS for that, like [here](https://tkn.me/tarantool-github-statistics-20200818/):

```css
table tr:nth-child(odd) td {
  background-color: #333;
}

table tr:nth-child(even) td {
  background-color: #555;
}
```

romanhabibov replied:
Done. I used it. See the last patch.

LeonidVas commented on templates/index.html:
```diff
@@ -0,0 +1,74 @@
+</head>
+<body>
+    <h1> Index of {{ path }} </h1>
+    <table id="indexlist">
```
Maybe we can make the code nicer like [there](https://stackoverflow.com/a/62838219)

romanhabibov replied:
Done. See the last patch.

romanhabibov commented:
> > > Please, remove the `.DS_Store` file.
> > > How icons you added are licensed and who is their author?
> > > Whether `index.html` is derived work (based on some other work)? If so, how it is licensed and who is the author?
> > 
> > 
> > +1
> > @romanhabibov ?
> 
> I think we should use something like [this one](https://google.github.io/material-design-icons/).

I used icons from https://icons.getbootstrap.com/ and added the license.

romanhabibov commented:
> Yet few comments / thoughts:
> 
> * We should not show the entire bucket, but start from the bucket/base/path (S3_BASE_PATH)
Done.

> * Maybe we want to have a filter for files that shouldn't be shown (what do you think @Totktonada)
???

> * Split the patch into commits
Done.

romanhabibov commented on s3repo/controller.py:
```diff
@@ -11,6 +11,7 @@
 from s3repo.model import S3ModelRequestError
 from s3repo.package import Package
 
+from s3repo.view import IndexView
```
Fixed.

romanhabibov commented on app.py:
```diff
@@ -71,9 +71,12 @@ def server_prepare():
 
     # Set the controller to work with S3.
     s3_view = S3Controller.as_view('s3_view', s3_model)
+    app.add_url_rule('/', view_func=s3_view,
```
See the patch about s3view. Did you mean it?

LeonidVas replied:
Yep.

romanhabibov commented on s3repo/controller.py:
```diff
@@ -91,10 +92,32 @@ def put(self, subpath):
+        view = IndexView()
+        err_msg = ''
+        try:
+            if type == 'folder':
```
I've dropped 'folder'.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,67 @@
+    files with browser.
+    """
+
+    def get_directory(self, subpath, items):
```
As I wrote earlier, it seems like "get_directory" and "get_file" are private methods.

romanhabibov replied:
Ok. Look at the current version.

LeonidVas commented:
Why do you ignore comments:
https://github.com/tarantool/rws/pull/18/commits/1522eb28a8d48b0f7e772f58a1cac53e8659e8a3#r652537052
https://github.com/tarantool/rws/pull/18/commits/1522eb28a8d48b0f7e772f58a1cac53e8659e8a3#r652689089

LeonidVas commented:
> > Yet few comments / thoughts:
> > 
> > * We should not show the entire bucket, but start from the bucket/base/path (S3_BASE_PATH)
> 
> Done.

No. I can still see the whole bucket
```
(rws) ┌─[leonid@vasya-hp]─[~/work/mail/rws]
└──╼ env | grep S3_BASE_PATH
S3_BASE_PATH=restored

```
![image](https://user-images.githubusercontent.com/57662928/124282530-9ac06280-db53-11eb-8a45-55083e48dbf0.png)

LeonidVas commented:
> > * Maybe we want to have a filter for files that shouldn't be shown (what do you think @Totktonada)
> 
> ???

For example (no legal) we only want to display files with ".deb and .rpm" extensions.

romanhabibov commented on s3repo/model.py:
```diff
@@ -2,8 +2,10 @@
 
 from multiprocessing.pool import ThreadPool
 import re
+import os
```
Dropped.

romanhabibov commented on app.py:
```diff
@@ -70,10 +70,9 @@ def server_prepare():
-    app.add_url_rule('/<path:subpath>', view_func=s3_view,
-        methods=['GET', 'PUT', 'DELETE'])
-
+    s3_controller = S3Controller.as_view('s3_controller', s3_model)
```
Modified the commit message.

romanhabibov commented on s3repo/model.py:
```diff
@@ -62,6 +65,8 @@ def __init__(self, s3_settings):
         )
         self.bucket = self.s3_resource.Bucket(self.s3_settings['bucket_name'])
 
+        self.s3_client = self.bucket.meta.client
```
Done.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
         NotImplementedError("delete_package hasn't been implemented yet.")
 
+    @staticmethod
+    def _list_items(objects, items):
```
Take a look at the current version.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
 
+    @staticmethod
+    def _list_items(objects, items):
+        """Process objects from list_objects_v2() and append they
```
Done.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+            contents = {}
+        for file_meta in contents:
+            file_name = file_meta.get('Key').split('/')[-1]
+            # list_objects() paginator returns the current
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+        if not objects.get('KeyCount'):
+            raise RuntimeError('No such folder.')
+
+        common_prefixes = objects.get('CommonPrefixes')
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+            item = Item('directory', directory_name)
+            items.append(item)
+
+        contents = objects.get('Contents')
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+
+        fields = ['Type', 'Name', 'LastModified', 'Size']
+        Item = namedtuple('Item', fields, defaults=['', '', '', ''])
+        if not objects.get('KeyCount'):
```
Added.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+            if not file_name:
+                continue
+            last_modified = file_meta.get('LastModified') \
+                           .strftime("%Y-%m-%d %H:%M:%S")
```
Fixed.

romanhabibov commented on app.py:
```diff
@@ -73,6 +74,14 @@ def server_prepare():
+    # Set the view to work with S3.
+    s3_view = S3View.as_view('s3_view', s3_model)
+    app.add_url_rule('/', view_func=s3_view,
+        methods=['GET'])
```
Fixed.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,67 @@
+"""View for working with the repositories on S3."""
+
+import os
+from collections import namedtuple
```
I do not understand what you mean.

LeonidVas replied:
I made a mistake while pasting the link.
https://github.com/tarantool/rws/pull/18#discussion_r652537052

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,67 @@
+from collections import namedtuple
+
+from flask.views import MethodView
+from flask import jsonify, render_template, Response
```
Done.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,67 @@
+    def __init__(self, model):
+        self.model = model
+
+    """S3View is just methods to display listings or download
```
Removed.

romanhabibov commented on static/icons/LICENSE.md:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Now, I have mentioned static/icons/LICENSE.md in LICENSE as Alexandr advised.

romanhabibov commented on s3repo/model.py:
```diff
@@ -2,8 +2,11 @@
+import os
 import subprocess as sp
 import time
+import datetime
```
Dropped

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+        """
+
+        fields = ['Type', 'Name', 'LastModified', 'Size']
+        Item = namedtuple('Item', fields, defaults=['', '', '', ''])
```
Done.

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+        try:
+            for objects in paginator.paginate(**list_parameters):
+                S3AsyncModel._list_items(objects, items)
+        except self.s3_client.exceptions.NoSuchBucket:
```
Removed. New issue https://github.com/tarantool/rws/issues/29

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +383,86 @@ def delete_package(self, package):
+        fields = ['Type', 'Name', 'LastModified', 'Size']
+        Item = namedtuple('Item', fields, defaults=['', '', '', ''])
+        if not objects.get('KeyCount'):
+            raise RuntimeError('No such folder.')
```
Done.

romanhabibov commented:
> > > Yet few comments / thoughts:
> > > 
> > > * We should not show the entire bucket, but start from the bucket/base/path (S3_BASE_PATH)
> > 
> > 
> > Done.
> 
> No. I can still see the whole bucket
> 
> ```
> (rws) ┌─[leonid@vasya-hp]─[~/work/mail/rws]
> └──╼ env | grep S3_BASE_PATH
> S3_BASE_PATH=restored
> ```
> 
> ![image](https://user-images.githubusercontent.com/57662928/124282530-9ac06280-db53-11eb-8a45-55083e48dbf0.png)

Done. See the current version.

romanhabibov commented:
> > > * Maybe we want to have a filter for files that shouldn't be shown (what do you think @Totktonada)
> > 
> > 
> > ???
> 
> For example (no legal) we only want to display files with ".deb and .rpm" extensions.

Let's do it as a separate issue.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+                err_msg = "Can't download file from S3: "
+                return S3View.get_file(abs_path, responce)
+            else:
+                return S3View.response_message('Wrong URL.', 404)
```
404 - [Not found](https://en.wikipedia.org/wiki/HTTP_404)

romanhabibov replied:
Added 404 html.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+            else:
+                return S3View.response_message('Wrong URL.', 404)
+        except RuntimeError as err:
+            return S3View.response_message(err_msg + str(err), 500)
```
I think the description of the server error does not need to be sent to the client.

romanhabibov replied:
Added 404 and 500 html.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+        displayed_path = os.path.normpath('/'.join(['', bucket_name, abs_path]))
+        parent_path = '/'.join(rel_path.split('/')[:-2])
+
+        # Make size human readable.
```
Can be moved to a separate method to make the code easier to read.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+    @staticmethod
+    def get_directory(abs_path, rel_path, bucket_name, items):
+        """Display directory content as a HTML page."""
+        displayed_path = os.path.normpath('/'.join(['', bucket_name, abs_path]))
```
Displayed path must not contain bucket name and base_path.

romanhabibov replied:
Fixed.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+        return render_template('index.html', **list_parameters)
+
+    @staticmethod
+    def get_file(path, responce):
```
responce -> response.
Here and bellow.

romanhabibov replied:
Fixed.

LeonidVas reviewed (changes requested):
From @Totktonada:
* CSS is uploaded every time (needs to be cached)
* CSS has a path with two slashes
* All fields must be left aligned
* The date field must be large enough to contain the date

romanhabibov commented:
> From @Totktonada:
> 
> * CSS is uploaded every time (needs to be cached)

added SEND_FILE_MAX_AGE_DEFAULT.
> * CSS has a path with two slashes

Fixed.

> * All fields must be left aligned

Done. In css file.

> * The date field must be large enough to contain the date

Now it contains.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+            headers={"Content-Disposition": "attachment; filename=" + filename}
+            )
+
+    def dispatch_request(self, subpath='/', type='directory'):
```
Done.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+        "subpath" path.
+        """
+        rel_path = '' if subpath.strip('/') == 'index' else subpath
+        base_path = self.model.s3_settings.get('base_path')
```
Done.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+        return response
+
+    @staticmethod
+    def get_directory(abs_path, rel_path, bucket_name, items):
```
I added "__".

LeonidVas replied:
No. One "_" is enough.

romanhabibov commented on s3repo/view.py:
```diff
@@ -0,0 +1,85 @@
+
+import os
+
+from flask.views import View
```
Fixed.

romanhabibov commented on app.py:
```diff
@@ -73,6 +74,13 @@ def server_prepare():
+    s3_view = S3View.as_view('s3_view', s3_model)
+    app.add_url_rule('/', view_func=s3_view, methods=['GET'])
+    app.add_url_rule('/<path:subpath>', view_func=s3_view, methods=['GET'])
+    app.add_url_rule('/<path:subpath>?<string:type>', view_func=s3_view,
```
Done. app.add_url_rule('/<path:subpath>?<string:type>' removed

romanhabibov commented on LICENSE:
```diff
@@ -28,3 +28,6 @@ LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.
 
+This license applies to all files except those in the static
```
Yes

romanhabibov commented on s3repo/model.py:
```diff
@@ -378,9 +437,49 @@ def delete_package(self, package):
     def get_file(self, path):
-        """Download a file from S3."""
-        NotImplementedError("get_file hasn't been implemented yet.")
+        """Get a file from S3 as a "Responce' object."""
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -143,6 +145,63 @@ def _format_paths(dist_path, dist_version, dist_base, filename):
+        common_prefixes = objects.get('CommonPrefixes') or {}
+        items = []
+        for prefix in common_prefixes:
+
```
Removed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -143,6 +145,63 @@ def _format_paths(dist_path, dist_version, dist_base, filename):
+        Item = namedtuple('Item', fields, defaults=['', '', '', ''])
+
+        # KeyCount is the number of keys returned with this
+        # request. KeyCount is 1 or emtpy directory (this folder
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -143,6 +145,63 @@ def _format_paths(dist_path, dist_version, dist_base, filename):
+        # +------+------+--------------+------+
+        Item = namedtuple('Item', fields, defaults=['', '', '', ''])
+
+        # KeyCount is the number of keys returned with this
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -143,6 +145,63 @@ def _format_paths(dist_path, dist_version, dist_base, filename):
+
+        fields = ['Type', 'Name', 'LastModified', 'Size']
+
+        # Item is tuple with metainfo of file or directory.
```
Fixed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -143,6 +145,63 @@ def _format_paths(dist_path, dist_version, dist_base, filename):
 
+    @staticmethod
+    def _list_items(objects):
+        """Process "objects" from list_objects_v2() and certain
```
Renamed.

romanhabibov commented on s3repo/model.py:
```diff
@@ -4,6 +4,7 @@
 import re
 import subprocess as sp
 import time
+from collections import namedtuple
```
Fixed.

LeonidVas commented:
> SEND_FILE_MAX_AGE_DEFAULT

LeonidVas commented:
> > * The date field must be large enough to contain the date
> 
> Now it contains.

![image](https://user-images.githubusercontent.com/57662928/126779935-6044adbd-bb3b-4f55-9362-0b2b7c3d5629.png)

Why don't you use one row for `date` and for `size` (it usually contains 2 rows now)

LeonidVas reviewed (commented):
Sorry I will reopen PR.

LeonidVas commented on s3repo/model.py:
```diff
@@ -1,5 +1,8 @@
 """Model for working with the repositories on S3."""
 
+import os
```
Why did you make "os" a separate group?

romanhabibov replied:
Fixed.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
+        type and absolute path. Type can be "directory", "file" or
+        None if object hasn't found."""
+
+        base_path = self.s3_settings.get('base_path') or ''
```
From `flake8`:
```
F841 local variable 'base_path' is assigned to but never used
```

romanhabibov replied:
Fixed.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
         return repos_list
 
+    @staticmethod
+    def _get_abs_path(base_path, path):
```
Don't make it `static` and get `base_path` from `self`.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
+    @staticmethod
+    def _get_abs_path(base_path, path):
+        """Get absolute (base_path + path) normalized path.
+        """
```
"""Get absolute (base_path + path) normalized path.""" - one line.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
+    def determine_type_and_path(self, path):
+        """Find an object spcified by "path" and determine its
+        type and absolute path. Type can be "directory", "file" or
+        None if object hasn't found."""
```
In the case of a multiline description, the last `"""` is placed on a separate line.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
+
+        base_path = self.s3_settings.get('base_path') or ''
+        abs_path = S3AsyncModel._get_abs_path(self.s3_settings.get('base_path'), path)
+        abs_path = abs_path.strip('/')
```
1) You can do it inside the `_get_abs_path` method.
2) This is the reason why I haven't used `os.path.normpath` before.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/model.py:
```diff
@@ -287,6 +346,49 @@ def _get_repository_list(self):
+        abs_path = abs_path.strip('/')
+
+        # Parameters for list_objects_v2():
+        # * 'Bucket' is a bucket name.
```
We usually used double quotes earlier in the description. Please replace it. Here and below.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,90 @@
+        self.model = model
+
+    @staticmethod
+    def response_message(message, status):
```
I think it might be private: `_response_message`

romanhabibov replied:
Removed.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,90 @@
+
+    @staticmethod
+    def __readable_size(size):
+        readable_size = ''
```
`__readable_size` -> `_readable_size`
Add a description.

romanhabibov replied:
Fixed.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,90 @@
+            if abs(size) < 1024.0:
+                readable_size = "%3.1f %sB" % (size, unit)
+                return readable_size
+            size /= 1024.0
```
Add some `return` in case of all iterations didn't go inside `if`.

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,90 @@
+            size /= 1024.0
+
+    @staticmethod
+    def __get_directory(rel_path, items):
```
`__get_directory` -> `_get_directory`

romanhabibov replied:
Done.

LeonidVas commented on s3repo/view.py:
```diff
@@ -0,0 +1,90 @@
+        return render_template('index.html', **list_parameters)
+
+    @staticmethod
+    def __get_file(path, response):
```
`__get_file` -> `_get_file`

romanhabibov replied:
Fixed.

Totktonada commented:
> Why don't you use one row for `date` and for `size` (it usually contains 2 rows now)?

I guess the reason is here:

```css
table {
<...>
    max-width: 700px;
<...>
}
```

LeonidVas commented on static/common/css/fileindex.css:
```diff
@@ -0,0 +1,70 @@
+    margin-top: 2em;
+    min-width: 400px;
+    text-align: center;
+    width: 100%;
```
With `width: 100%;`
![CSS_1](https://user-images.githubusercontent.com/57662928/126954550-3e1bcac6-1ef3-4647-8927-9c32108b6354.png)
Without `width: 100%;`
![CSS_2](https://user-images.githubusercontent.com/57662928/126954648-42336025-7fbc-4331-9507-decbb75488d7.png)
As for me, the second is prettier.

romanhabibov replied:
Second. Fixed.

LeonidVas reviewed (changes requested):
I think we are already close to the finish

romanhabibov commented on s3repo/model.py:
```diff
@@ -402,9 +501,50 @@ def delete_package(self, package):
     def get_file(self, path):
-        """Download a file from S3."""
-        NotImplementedError("get_file hasn't been implemented yet.")
+        """Get a file from S3 as a boto3 get_object() object."""
```
Fixed.

LeonidVas commented:
@romanhabibov I suggest making several fixups if you are ok, please, squash them in to your commits.

LeonidVas reviewed (approved):
LGTM.
